File: 백준/Gold/2457. 공주님의 정원/공주님의 정원.py
import sys

# 꽃들의 총 개수 N
n = int(sys.stdin.readline())

# 꽃들이 피고 지는 날짜
dates = []
for _ in range(n):
    # 꽃이 피고 지는 날짜에 대해 월에 100을 곱한 뒤, 일이랑 덧셈
    start_m, start_d, end_m, end_d = map(int, sys.stdin.readline().split())
    dates.append([start_m * 100 + start_d, end_m * 100 + end_d])
dates.sort() # 오름차순 정렬

end = 301
cnt = 0
while dates:
    # 첫 시작날짜가 301보다 크면 0 출력해야함
    # 저번 반복의 마지막 끝날짜 > 1130면 cnt만 보면됌 
    #  끝날짜 < 첫 시작날짜면 (앞조건에서 걸렀으므로)비는 날이 생기는 경우만 발생
    if end >= 1201 or dates[0][0] > end:
        break
    
    # end보다 일찍 피는 꽃들 중에서 가장 늦게 지는 꽃 찾기
    temp_end = -1
    for _ in range(len(dates)):
        if dates[0][0] <= end:
            if temp_end <= dates[0][1]:
                temp_end = dates[0][1] # 이전 end보다 크거나 같은 것만 저장
            
            dates.remove(dates[0])
        else:
            break
    
    end = temp_end
    cnt += 1
            
if end < 1201:
    print(0)
else:
    print(cnt)

File: 백준/Gold/2457. 공주님의 정원/공주님의 정원.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    private void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine().trim());
        List<List<Integer>> dates = new ArrayList<>();
        for (int i=0; i < n; i++) {
            String[] input = br.readLine().split(" ");
            List<Integer> date = List.of(
                    Integer.parseInt(input[0]) * 100 + Integer.parseInt(input[1]),
                    Integer.parseInt(input[2]) * 100 + Integer.parseInt(input[3]));
            dates.add(date);
        }
        dates.sort(Comparator.comparing((List<Integer> list) -> list.get(0)));


        int end = 301; // 마지막 꽃의 지는 날짜
        int cnt = 0;
        while (!dates.isEmpty()) {
            if (dates.get(0).get(0) > end || end > 1130) { // 심을 꽃이 없거나(초기상태) 마지막 꽃의 지는 날짜가 지금 꽃의 피는 날짜보다 크거나 1130보다 크면 멈춤
                break;
            }

            int subEnd = 0;
            while (!dates.isEmpty()) {
                if (dates.get(0).get(0) <= end) { // 피는 날짜가 마지막 꽃의 지는 날짜보다 작거나 같으면 가장 큰 지는 날짜 구함
                    subEnd = Math.max(dates.get(0).get(1), subEnd);
                    dates.remove(0);
                } else { // 피는 날짜가 마지막 꽃의 지는 날짜보다 크면 멈춤
                    break;
                }
            }

            end = subEnd; // 가장 큰 지는 날짜가 마지막 꽃의 지는 날짜가 되거나 꽂이 필수 없는 기간이 생기면 0이됨
            cnt++;
        }

        if (end < 1201) {
            System.out.println(0);
        } else {
            System.out.println(cnt);
        }
    }

    public static void main(String[] args) throws Exception {
        new Main().solution();
    }
}
